src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/RawConnectionStream.cs

```diff
@@ -175,7 +176,7 @@ public override ValueTask WriteAsync(ReadOnlyMemory<byte> buffer, CancellationTo
                 HttpConnection connection = _connection;
                 if (connection == null)
                 {
-                    return new ValueTask(Task.FromException(new IOException(SR.ObjectDisposed_StreamClosed)));
+                    return new ValueTask(Task.FromException(ExceptionDispatchInfo.SetCurrentStackTrace(new IOException(SR.ObjectDisposed_StreamClosed))));
```

Comment:
`Task.FromException(ExceptionDispatchInfo.SetCurrentStackTrace` looks like a common pattern.

Would it make sense for `Task.FromException` to set the stack trace if it was not set already?

Reply:
It would make sense; same for {Try}SetException. My hesitancy is there will be a race condition if the same exception object is used concurrently with multiple such calls and/or directly using SetCurrentStackTrace (there's obviously already a race if the latter is used directly concurrently, but this would hide that a bit). You think that's ok?

Reply:
I updated the coreclr PR to set this in Task.FromException and TCS.TrySetException, and then reverted most of the usage in this PR.

Reply:
We have number of corner case situations like this today where the same exception instance can be thrown on multiple threads at the same time. You may get completely bogus stacktrace attached to the exception when this happens. It does not seem to be problem in practice. I do not recall any customer reports of bogus stacktraces.

The root cause of the problem is that there is no good way to clone an exception. The problem was discussed when ExceptionDispatchInfo was being designed and we have decided to just live with it.

I would be more worried about the performance impact of attaching formatted stack to the exceptions eagerly. Formatting the stack is an expensive operation. We are careful to format the stacktrace lazily, but the current implementation does not follow the lazy plan. If somebody is using exceptions for control flow, this change may have quite noticeable performance impact.

Reply:
> I would be more worried about the performance impact of attaching formatted stack to the exceptions eagerly.

That is potentially concerning.  What would you recommend?  Reverting the change that included it automatically in Task.FromException/TrySetException?  Or is there an efficient way I could populate _stackTrace, akin to when the exception is thrown but with the full stack trace from the current location?  Or leaving it and seeing what happens?

Reply:
> is there an efficient way I could populate _stackTrace, akin to when the exception is thrown

It is doable, but it requires changes in the unmanaged runtime portion.

> Or leaving it and seeing what happens?

I think we can leave this and see what happens.

Reply:
> I think we can leave this and see what happens.

Hmm.  I just did some quick microbenchmarks.  This has no measurable impact on `try { ... } catch (Exception e) { tcs.SetException(e); }` as you'd expect, since in that case we're not actually generating the textual stack trace.  But, it makes `tcs.SetException(new Exception())` ~60x slower, which is also ~4x slower than the throw/catch.  That's really unfortunate, and I worry it's large enough to matter.

Reply:
It looks like the `new StackTrace(true)` is half the cost (~35us), and the `ToString` is half the cost (~30us).

Also, my methodology was slightly flawed: I was comparing an exception that had to walk effectively 0 frames with effectively `new StackTrace(true)`, which had to walk (in this case based on how Benchmark.NET was calling it) 12 frames.  When I make the exception pass through 12 frames as well (via a bunch of NoInlining helpers), it takes ~25us, so ~30% faster than `new StackTrace(true)`.

Back of the napkin, that suggests that if we were to do the optimal thing here, making runtime changes to enable just doing the stack walk and delaying the string until later, the best we could hope for would be, e.g. ~25us instead of ~60us.  Certainly faster, but not earth shattering.  There's also the managed garbage involved in `new StackTrace(true).ToString()`, of course.

At this point, I'm thinking we should go back to what I had previously: expose the API, use it in various places in corefx, but not automatically use it in FromException / TrySetException.  Library authors can opt-in to using it then.

Reply:
Sounds good to me.